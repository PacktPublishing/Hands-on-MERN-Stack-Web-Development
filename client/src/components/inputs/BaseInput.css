.BaseInput {
    display: flex;
    flex-direction: column;
}

.BaseInputReverse {
    flex-direction: row-reverse;
    align-content: flex-start;
}

.BaseInputReverse label {
    flex: 1;
    padding-left: 0.75rem;
}

.BaseInput input {
    display: flex;
    padding: 0.75rem;
    margin-bottom: 0.75rem;
    font-family: "Courier New", monospace;
    border: none;
    border-bottom: 2px solid antiquewhite;
    transition: 220ms all;
    outline: 2px solid transparent;

}

.BaseInput input:focus {
    outline: 2px solid aquamarine;
}

.BaseInput label {
    display: flex;
    font-size: 0.85rem;
    font-family: Monaco, monospace;
    cursor: pointer;
}