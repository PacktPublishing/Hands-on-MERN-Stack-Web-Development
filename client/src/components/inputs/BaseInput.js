import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './BaseInput.css';

export default class BaseInput extends Component {
  render() {
    return (
      <div className={
        `BaseInput ${this.props.type === 'checkbox' ? 'BaseInputReverse' : ''}`
      }>
        <label htmlFor={this.props.name} data-testid="baseInputLabel">
          {this.props.label}
        </label>
        <input
          data-testid="baseInput"
          id={this.props.name}
          {...this.props}
        />
      </div>
    )
  }
}

BaseInput.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  type: PropTypes.oneOf(['checkbox', 'text', 'password']).isRequired,
};
