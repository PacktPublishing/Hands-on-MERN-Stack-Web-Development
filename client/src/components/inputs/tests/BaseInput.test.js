import React from 'react';
import { render, fireEvent, cleanup } from 'react-testing-library';
import renderer from 'react-test-renderer';
import BaseInput from '../BaseInput';

const mockCallback = jest.fn(e => e);

const setup = () => {
  const utils = render(
    <BaseInput
      name="firstName"
      label="First Name"
      type="text"
      onChange={mockCallback}
    />
  );
  const input = utils.getByTestId('baseInput');
  const label = utils.getByTestId('baseInputLabel');
  return {
    input,
    label,
    ...utils,
  };
};

afterEach(cleanup);

test('Should call mockCallback on text input', () => {
  const { input } = setup();
  fireEvent.change(input, { target: { value: 'Hello!' } });
  fireEvent.change(input, { target: { value: 'World!' } });
  expect(mockCallback.mock.calls.length).toBe(2);
});

test('Should contain proper inner html and htmlFor attribute', () => {
  const { label } = setup();
  expect(label.innerHTML).toBe('First Name');
  expect(label.htmlFor).toBe('firstName');
});

test('Should match snapshot', () => {
  const output = renderer.create(
    <BaseInput
      name="firstName"
      label="First Name"
      type="text"
      onChange={() => {}}
    />
  );
  expect(output.toJSON()).toMatchSnapshot();
});
