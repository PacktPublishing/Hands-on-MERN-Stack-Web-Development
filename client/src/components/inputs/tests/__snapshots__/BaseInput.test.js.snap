// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should match snapshot 1`] = `
<div
  className="BaseInput "
>
  <label
    data-testid="baseInputLabel"
    htmlFor="firstName"
  >
    First Name
  </label>
  <input
    data-testid="baseInput"
    id="firstName"
    label="First Name"
    name="firstName"
    onChange={[Function]}
    type="text"
  />
</div>
`;
